src/index.js

```diff
@@ -0,0 +1,135 @@
+'use strict'
+
+const { NOISE } = require('@chainsafe/libp2p-noise')
+const libp2p = require('libp2p')
+const Multiplex = require('libp2p-mplex')
+const Websockets = require('libp2p-websockets')
+const { blocksTable, getPeerId, primaryKeys, port } = require('./config')
+const { logger, serializeError } = require('./logging')
+const {
+  Message,
+  BITSWAP_V_100,
+  BITSWAP_V_110,
+  BITSWAP_V_120,
+  Block,
+  BlockPresence,
+  Entry,
+  WantList
+} = require('./message')
+const { receiveData, sendData } = require('./networking')
+const { cidToKey, fetchS3Object, readDynamoItem } = require('./storage')
+
+const emptyWantList = new WantList([], true)
+
+async function fetchBlock(cid) {
+  const info = await getBlockInfo(cid)
+
+  if (!info) {
+    return null
+  }
+
+  const { offset, length, car } = info.cars[0]
+  const separator = car.indexOf('/')
+  const bucket = car.slice(0, separator)
+  const key = car.slice(separator + 1)
+
+  return fetchS3Object(bucket, key, offset, length)
+}
+
+async function getBlockInfo(cid) {
+  const item = await readDynamoItem(blocksTable, primaryKeys.blocks, cidToKey(cid))
+  return item
+}
+
+async function processWantlist(stream, protocol, wantlist) {
+  const blocks = []
+  const presences = []
+
+  // For each entry in the list
+  for (const entry of wantlist.entries) {
```

Comment:
I would recommend to run this in parallel.

Reply:
I agree.
I initially wrote it serially to ensure compatibility. I've asked PL people if we can run this in parallel.

Reply:
Rewritten using `p-all`. It will be in the next push.